bitcoin/src/address/mod.rs

```diff
@@ -514,32 +570,34 @@ impl Address {
 
     /// Constructs an [`Address`] from an output script (`scriptPubkey`).
     pub fn from_script(script: &Script, network: Network) -> Result<Address, Error> {
-        let payload = if script.is_p2pkh() {
+        if script.is_p2pkh() {
             let bytes = script.as_bytes()[3..23].try_into().expect("statically 20B long");
-            Payload::PubkeyHash(PubkeyHash::from_byte_array(bytes))
+            let hash = PubkeyHash::from_byte_array(bytes);
+            Ok(Address::_p2pkh(hash, network))
         } else if script.is_p2sh() {
             let bytes = script.as_bytes()[2..22].try_into().expect("statically 20B long");
-            Payload::ScriptHash(ScriptHash::from_byte_array(bytes))
+            let hash = ScriptHash::from_byte_array(bytes);
+            Ok(Address::_p2sh(hash, network))
         } else if script.is_witness_program() {
-            let opcode = script.first_opcode().expect("witness_version guarantees len() > 4");
-
-            let witness_program = script.as_bytes()[2..].to_vec();
+            let opcode = script.first_opcode().expect("is_witness_program guarantees len > 4");
 
-            let witness_program =
-                WitnessProgram::new(WitnessVersion::try_from(opcode)?, witness_program)?;
-            Payload::WitnessProgram(witness_program)
+            let buf = PushBytesBuf::try_from(script.as_bytes()[2..].to_vec())
```

Comment:
Oh, I realized I forgot to implement `TryFrom<&[u8]> for PushBytesBuf` which would make this not allocate in case of error. Will do a followup PR after this one. That is causing all the noise in the code now. Although `<&PushBytes>::try_from(...).expect("...").to_owned()` works too for perf (but still causes noise).